apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -0,0 +1,259 @@
+
+    // AUTHENTICATE
+    await authnLib.loadUser(req, res, userInfo);
+
+    // Record the LTI 1.3 user's subject id
+    await queryAsync(sql.update_lti13_users, {
+      user_id: res.locals.authn_user.user_id,
+      lti13_instance_id: lti13_instance.id,
+      sub: lti13_claims.sub,
+    });
+
+    // Get the target_link out of the LTI request and redirect
+    const redirUrl =
+      lti13_claims['https://purl.imsglobal.org/spec/lti/claim/target_link_uri'] ||
+      '/pl';
+    res.redirect(redirUrl);
+  }),
+);
+
+export default router;
+
+//
+// Schema to validate OIDC, LTI
+//
+const OIDCAuthResponseSchema = z.object({
+  state: z.string(),
+  id_token: z.string(),
+});
+
+const OIDCLaunchFlowSchema = z.object({
+  iss: z.string(),
+  login_hint: z.string(),
+  target_link_uri: z.string(),
+});
+
+// Validate LTI 1.3
+// https://www.imsglobal.org/spec/lti/v1p3#required-message-claims
+const LTI13Schema = z.object({
+  'https://purl.imsglobal.org/spec/lti/claim/message_type': z.literal('LtiResourceLinkRequest'),
+  'https://purl.imsglobal.org/spec/lti/claim/version': z.literal('1.3.0'),
+  'https://purl.imsglobal.org/spec/lti/claim/deployment_id': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/resource_link': z.object({
+    id: z.string(),
+    description: z.string().nullable(),
+    title: z.string().nullable(),
+  }),
+  sub: z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/roles': z.string().array(),
+});
+
+//
+// Helper functions
+//
+
+async function authenticate(req: Request, res: Response): Promise<any> {
+  // https://www.imsglobal.org/spec/security/v1p0/#step-3-authentication-response
+  OIDCAuthResponseSchema.parse(req.body);
+
+  const myPassport = await setupPassport(req.params.lti13_instance_id);
+  return new Promise((resolve, reject) => {
+
+    // Callback arguments described at
+    // https://github.com/jaredhanson/passport/blob/33b92f96616642864844753a481df7c5b823e047/lib/middleware/authenticate.js#L34
+    myPassport.authenticate(`lti13`, ((err, user, info) => {
+      if (err) {
+        reject(err);
+      } else if (!user) {
+        // The authentication libraries under openid-connect will fail (silently) if the key length
+        // is too small, like with the Canvas development keys. It triggers that error in PL here.
+        reject(
+          error.make(400, `Authentication failed, before user validation.`, {
+            err,
+            user,
+            info_raw: info,
+            info: info?.toString(),
+          }),
+        );
+      } else {
+        resolve(user);
+      }
+    }) as passport.AuthenticateCallback)(req, res);
+  });
+}
+
+async function launchFlow(req: Request, res: Response, next: NextFunction) {
+  // https://www.imsglobal.org/spec/security/v1p0/#step-1-third-party-initiated-login
+
+  const parameters = { ...req.body, ...req.query };
+
+  OIDCLaunchFlowSchema.parse(parameters);
```

Comment:
Sorry, missed this on my last review.

```suggestion
  const parameters = OIDCLaunchFlowSchema.parse({ ...req.body, ...req.query });
```